### PR TITLE
Fix #825

### LINKED ISSUE #825: Using `abstract` keyword on abstract method returning inner type causes compiler crash
Compiling this:
```
trait X {
  type Y
  abstract def y: Y
}

object Z {
  val x: X = new X {
    type Y = Int
    def y: Int = 1
  }  
}
```
causes the following compiler crash during bytecode generation:
```
java.lang.NullPointerException
        at scala.tools.asm.tree.MethodNode.visitLocalVariable(MethodNode.java:571)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:625)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:641)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:510)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$$anonfun$gen$1.apply(BCodeSkelBuilder.scala:512)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$$anonfun$gen$1.apply(BCodeSkelBuilder.scala:512)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:512)
        at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:119)
        at dotty.tools.backend.jvm.GenBCodePipeline$Worker1.visit(GenBCode.scala:175)
        at dotty.tools.backend.jvm.GenBCodePipeline$Worker1.run(GenBCode.scala:127)
        at dotty.tools.backend.jvm.GenBCodePipeline.buildAndSendToDisk(GenBCode.scala:329)
        at dotty.tools.backend.jvm.GenBCodePipeline.run(GenBCode.scala:298)
        at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:41)
        at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:271)
        at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:269)
        at scala.collection.immutable.List.map(List.scala:273)
        at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:269)
        at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:34)
        at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:59)
        at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:56)
        at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
        at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:56)
        at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:52)
        at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:52)
        at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:69)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:52)
        at dotty.tools.dotc.Run.compileSources(Run.scala:49)
        at dotty.tools.dotc.Run.compile(Run.scala:33)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:21)
        at dotty.tools.dotc.Driver.process(Driver.scala:44)
        at dotty.tools.dotc.Driver.main(Driver.scala:48)
        at dotty.tools.dotc.Main.main(Main.scala)
error: Error while emitting test.scala
```

However, removing the `abstract` keyword results in successful compilation. It seems like the `abstract` keyword, when applied to an already-abstract method, should have no effect on the output.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #825 as one of many well-formedness conditions. Review by @smarter.

### CODE REVIEW

**Comment by smarter** [comment]
Otherwise LGTM.

File: src/dotty/tools/dotc/ast/Desugar.scala [hunk 1]
```diff
@@ -380,18 +380,22 @@ object desugar {
...
-        if (mods is Case)
-          ctx.error("implicit classes may not case classes", cdef.pos)
-
+      if (!mods.is(Implicit))
+        Nil
+      else if (ctx.owner is Package) {
+        ctx.error("implicit classes may not be toplevel", cdef.pos)
+        Nil
+      }
+      else if (mods is Case) {
```
Comment (smarter) [inline src/dotty/tools/dotc/ast/Desugar.scala]: I think our coding style is `} else if ...` without a line break after the `}`.

Comment (odersky) [inline src/dotty/tools/dotc/ast/Desugar.scala]: No I write newline } quite systematically now. Makes keywords align better.


On Wed, Oct 21, 2015 at 6:43 PM, Guillaume Martres <notifications@github.com
> wrote:

> In src/dotty/tools/dotc/ast/Desugar.scala
> <https://github.com/lampepfl/dotty/pull/834#discussion_r42649286>:
>
> > @@ -380,18 +380,22 @@ object desugar {
> >      //     synthetic implicit C[Ts](p11: T11, ..., p1N: T1N) ... (pM1: TM1, ..., pMN: TMN): C[Ts] =
> >      //       new C[Ts](p11, ..., p1N) ... (pM1, ..., pMN) =
> >      val implicitWrappers =
> > -      if (mods is Implicit) {
> > -        if (ctx.owner is Package)
> > -          ctx.error("implicit classes may not be toplevel", cdef.pos)
> > -        if (mods is Case)
> > -          ctx.error("implicit classes may not case classes", cdef.pos)
> > -
> > +      if (!mods.is(Implicit))
> > +        Nil
> > +      else if (ctx.owner is Package) {
> > +        ctx.error("implicit classes may not be toplevel", cdef.pos)
> > +        Nil
> > +      }
> > +      else if (mods is Case) {
>
> I think our coding style is } else if ... without a line break after the }
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/834/files#r42649286>.
>



-- 
Martin Odersky
EPFL

**Comment by odersky** [comment]
Fixed reviewers's comments and rebased to master.

**Pull request merged**